File: Plugin/Source/S950/Engine.cpp
#include "Engine.h"

#include <cstring>

namespace s950
{
    Engine::Engine (double rate)
        : sampleRate (rate <= 0 ? 48000.0 : rate)
    {
        matched.reserve (Polyphony);
    }

    /*
     * WHY THE PATCH CHANGES HANDS THIS WAY
     *
     * The C# simply assigned a reference and let the garbage collector work out when the old
     * programme could go. There is no collector here, and the two obvious replacements are
     * both wrong in the same place:
     *
     *   - An atomic<Patch*> leaks, or frees while a voice is halfway through a sample.
     *   - A shared_ptr read on the audio thread copies, which touches a reference count and,
     *     if that copy turns out to be the last, frees every sample buffer in the programme
     *     inside the audio callback. That is a dropout waiting for the worst moment.
     *
     * So the pointer is moved, never copied, and only ever in one direction at a time. The
     * message thread puts a patch in `pending`; the audio thread moves it into `audioPatch`
     * and moves what was there into `retired`; the message thread takes `retired` away and
     * destroys it. Moving a shared_ptr touches no reference count and allocates nothing, so
     * the audio thread's half costs two pointer swaps.
     *
     * The audio thread declines to take a new patch while a retired one is still waiting to
     * be collected. That means a second programme change before the host's message thread
     * has run is simply held until it has - a programme arriving a few milliseconds late
     * being the whole cost of never freeing on the wrong thread.
     */
    void Engine::setPatch (PatchPtr patch)
    {
        collectRetiredPatch();          // make room, if the audio thread has handed one back

        // The shared LFO runs at the rate the programme's keygroups ask for. They almost
        // always agree; where they do not, the first one wins, since one shared oscillator
        // cannot be at two rates at once.
        double step = 0.0;
        if (patch != nullptr)
        {
            for (const auto& k : patch->keygroups)
            {
                if (k.lfoDesync) continue;

                step = 2.0 * 3.14159265358979323846
                       * (cal::LfoRateHzAtZero + k.lfoRate * cal::LfoRateHzPerUnit) / sampleRate;
                break;
            }
        }

        sharedStep = step;              // read only by the audio thread; a double write is atomic enough here
        pending    = std::move (patch);
        pendingReady.store (true, std::memory_order_release);
    }

    void Engine::collectRetiredPatch()
    {
        if (! retiredReady.load (std::memory_order_acquire))
            return;

        retired.reset();                // the frees happen HERE, on the message thread
        retiredReady.store (false, std::memory_order_release);
    }

    void Engine::takePendingPatch()
    {
        if (! pendingReady.load (std::memory_order_acquire))
            return;

        // Only if the last one has been collected - otherwise we would have nowhere to put
        // what we are replacing, and dropping it here would free it on this thread.
        if (retiredReady.load (std::memory_order_acquire))
            return;

        retired    = std::move (audioPatch);
        audioPatch = std::move (pending);

        pendingReady.store (false, std::memory_order_release);
        retiredReady.store (true,  std::memory_order_release);

        // Notes already sounding take up the new settings where they stand, rather than
        // waiting to be struck again.
        repatch();
    }

    // ---------------------------------------------------------------------- the ring

    void Engine::post (unsigned char kind, int a, int b, int at)
    {
        const int w    = writeIndex.load (std::memory_order_relaxed);
        const int next = (w + 1) % RingSize;

        if (next == readIndex.load (std::memory_order_acquire))
            return;                     // full: drop it rather than block

        ring[w].kind = kind;
        ring[w].a    = static_cast<unsigned char> (a < 0 ? 0 : (a > 255 ? 255 : a));
        ring[w].b    = static_cast<unsigned char> (b < 0 ? 0 : (b > 255 ? 255 : b));
        ring[w].at   = static_cast<unsigned short> (at < 0 ? 0 : (at > 65535 ? 65535 : at));

        writeIndex.store (next, std::memory_order_release);
    }

    bool Engine::peekEvent (int& at, int count) const
    {
        const int r = readIndex.load (std::memory_order_relaxed);

        if (r == writeIndex.load (std::memory_order_acquire))
            return false;

        /*
         * Clamped into this block rather than carried over to the next.
         *
         * A host should never hand over an offset past the end of the block it came
         * with, but if one does, holding the event back would mean keeping state about
         * a block that has already gone. Late by a few samples beats lost.
         */
        at = ring[r].at;
        if (at > count - 1) at = count - 1;
        if (at < 0)         at = 0;

        return true;
    }

    void Engine::applyNextEvent()
    {
        const int r = readIndex.load (std::memory_order_relaxed);

        if (r == writeIndex.load (std::memory_order_acquire))
            return;

        const Event e = ring[r];
        readIndex.store ((r + 1) % RingSize, std::memory_order_release);

        switch (e.kind)
        {
            case EvNoteOn:  startNote (e.a, e.b); break;
            case EvNoteOff: stopNote (e.a);       break;
            case EvWheel:   wheel = e.a;                break;
            case EvPressure: pressure = e.a;            break;
            case EvBend:    bend14 = (e.a << 7) | e.b;  break;

            case EvAllOff:
                for (auto& v : voices) v.release();
                break;

            default: break;
        }
    }

    /*
     * Hand every sounding voice its keygroup out of the patch now loaded.
     *
     * Matched by keygroup index rather than by key range: a voice belongs to the keygroup
     * that started it, and an edit may have moved the ranges under it.
     */
    void Engine::repatch()
    {
        if (audioPatch == nullptr)
        {
            // Nothing to point at any more. The voices hold a shared_ptr to their audio, so
            // they can finish what they are playing - but their keygroup is gone.
            for (auto& v : voices)
                if (v.isActive()) v.kill();

            return;
        }

        for (auto& v : voices)
        {
            if (! v.isActive()) continue;

            const int want = v.getKeygroupIndex();
            if (want < 0) continue;

            for (const auto& kg : audioPatch->keygroups)
            {
                if (kg.keygroupIndex != want) continue;
                if (v.getVelocity() < kg.velocityFrom || v.getVelocity() > kg.velocityTo) continue;

                v.adopt (kg);
                break;
            }
        }
    }

    void Engine::startNote (int note, int velocity)
    {
        if (audioPatch == nullptr)
            return;

        audioPatch->matching (note, velocity, matched);

        /*
         * The positional crossfade needs every keygroup answering this note at once, so the
         * ranges are gathered before any voice starts. With the flag off, or with one
         * keygroup answering, crossfadeGain returns 1 and nothing below changes.
         *
         * Fixed arrays rather than a vector: this is the audio thread, and one note can
         * start no more voices than the polyphony allows.
         */
        int fadeLow[Polyphony], fadeHigh[Polyphony];
        const int fadeCount = std::min (static_cast<int> (matched.size()), Polyphony);

        for (int m = 0; m < fadeCount; ++m)
        {
            fadeLow[m]  = matched[static_cast<size_t> (m)]->lowKey;
            fadeHigh[m] = matched[static_cast<size_t> (m)]->highKey;
        }

        int index = -1;

        for (const KeygroupPatch* kg : matched)
        {
            ++index;

            const double fade = (audioPatch->positionalCrossfade && index < fadeCount)
                                  ? cal::crossfadeGain (note, fadeLow, fadeHigh,
                                                        fadeCount, index)
                                  : 1.0;

            /*
             * What the two performance controllers add, in cents.
             *
             * The modwheel is byte 22 and channel pressure is byte 21, and the aftertouch
             * run measured them to be THE SAME MECHANISM from different sources. At full,
             * aftertouch gave 71.95 cents against the wheel's 72.3 - one constant, not
             * two. Byte 21 scales it proportionally, reading 0.511 of full at 50 where a
             * straight proportion is 0.505 and byte 22 gave 0.509. And the two ADD: wheel
             * alone read 71.87 cents, wheel and pressure together 149.79, where taking the
             * larger would have left it at 71.87.
             *
             * Byte 21 used to be read and dropped, on the grounds that it is 0 in all 1908
             * keygroups of one person's disks. Other people have other disks.
             */
            double wheelCents = cal::LfoWheelCentsAtFull
                                  * (kg->lfoModwheelDepth / 99.0)
                                  * (wheel / 127.0)
                              + cal::LfoWheelCentsAtFull
                                  * (kg->lfoAftertouchDepth / 99.0)
                                  * (pressure / 127.0);

            if (kg->lfoDepth * cal::LfoDepthCentsPerUnit + wheelCents < 0.5)
                wheelCents = 0.0;

            /*
             * The trims go on before start(), not after: start() works out the envelope from
             * them and primes the filter with the cutoff at time zero, so a voice that began
             * from untrimmed settings would attack wrongly and click its way to the right
             * ones.
             */
            Voice& v = take();
            v.setTrims (trims.read());
            v.start (*kg, note, velocity, sampleRate, wheelCents, sequence++, fade);
        }
    }

    /*
     * Let go of every voice holding this note.
     *
     * Every one, not the first: pressing the same key twice before releasing it is ordinary,
     * and leaving the older voice held would strand it.
     */
    void Engine::stopNote (int note)
    {
        for (auto& v : voices)
            if (v.isActive() && v.getNote() == note && v.isHeld())
                v.release();
    }

    /*
     * A voice to start a note on.
     *
     * Idle first, obviously. Then the oldest one that has already been let go and is only
     * fading - taking that costs a tail nobody is listening to. Only if every voice is still
     * held does it take one of those, and then the oldest.
     *
     * Taking the oldest regardless, which the C# did at first, would cut off a key the player
     * is holding while a released note was left ringing beside it - which reads exactly like
     * "the release did not happen".
     */
    Voice& Engine::take()
    {
        for (auto& v : voices)
            if (! v.isActive()) return v;

        int pick = -1;
        for (int i = 0; i < Polyphony; ++i)
        {
            if (voices[i].isHeld()) continue;                 // still down: leave it
            if (pick < 0 || voices[i].getStartedAt() < voices[pick].getStartedAt()) pick = i;
        }

        if (pick < 0)
        {
            pick = 0;
            for (int i = 1; i < Polyphony; ++i)
                if (voices[i].getStartedAt() < voices[pick].getStartedAt()) pick = i;
        }

        voices[pick].kill();
        return voices[pick];
    }

    int Engine::getActiveVoices() const
    {
        int n = 0;
        for (const auto& v : voices)
            if (v.isActive()) ++n;

        return n;
    }

    void Engine::renderSpan (float* left, float* right, int count)
    {
        if (count <= 0) return;

        // Read once for the whole stretch, so every voice in it is shaped by the same
        // settings and a control moved mid-block cannot land differently on two voices.
        const Trims now = trims.read();

        /*
         * The pitch wheel, as a multiplier on the playback rate, worked out once per stretch.
         *
         * Read here rather than at note-on because a bend has to reach notes that are already
         * sounding - that is the whole point of a wheel, and it is what separates it from
         * velocity, which is settled when the key goes down.
         *
         * 8192 is the rest position and the two halves are not the same width: 8192 steps
         * below it and 8191 above. Dividing by 8192 either way would make a full upward bend
         * fall one step short of the range, which is inaudible but wrong; dividing by the
         * right half of the range gets both ends exactly.
         */
        const double bendNow =
            cal::bendRatio (bend14, bendRange.load (std::memory_order_relaxed));

        for (auto& v : voices)
            if (v.isActive())
            {
                v.setTrims (now);
                v.setBend (bendNow);
                v.render (left, right, count, sharedPhase);
            }

        // The shared LFO moves with the audio, so it advances per stretch rather than
        // once per block - otherwise splitting a block would change how it sounds.
        sharedPhase += sharedStep * count;
        if (sharedPhase > 2.0 * 3.14159265358979323846)
            sharedPhase = std::fmod (sharedPhase, 2.0 * 3.14159265358979323846);
    }

    /*
     * Fill the block, stopping at each event to do it where it belongs.
     *
     * The C# renders a block and then applies whatever arrived, because a keyboard and a
     * MIDI port have no finer timing to give it. A host does: every note comes with an
     * offset into the block it was handed with, and rounding those to the block boundary
     * is up to 11 ms of jitter at a 512-sample buffer.
     *
     * So the block is rendered in stretches between events. With nothing in the ring
     * that is one stretch and the same work as before; with a note at sample 200 of 512
     * it is two, and the note starts on sample 200.
     */
    void Engine::render (float* buffer, float* right, int count)
    {
        takePendingPatch();

        if (count <= 0) return;

        std::memset (buffer, 0, static_cast<size_t> (count) * sizeof (float));
        if (right != nullptr)
            std::memset (right, 0, static_cast<size_t> (count) * sizeof (float));

        int at = 0;
        while (at < count)
        {
            int next;

            // everything due by now, in the order it arrived
            while (peekEvent (next, count) && next <= at)
                applyNextEvent();

            // up to the next one, or to the end of the block
            int until = peekEvent (next, count) ? next : count;
            if (until <= at)  until = at + 1;      // never stand still
            if (until > count) until = count;

            renderSpan (buffer + at, right != nullptr ? right + at : nullptr, until - at);
            at = until;
        }

        const float g = gain.load (std::memory_order_relaxed);

        for (int i = 0; i < count; ++i)
        {
            const float v = buffer[i] * g;
            buffer[i] = v > 1.0f ? 1.0f : (v < -1.0f ? -1.0f : v);

            if (right != nullptr)
            {
                const float r = right[i] * g;
                right[i] = r > 1.0f ? 1.0f : (r < -1.0f ? -1.0f : r);
            }
        }
    }
}

File: AkaiS950Engine/Engine.cs
using System;
using System.Collections.Generic;

namespace AkaiS950Engine
{
    /// <summary>
    /// The instrument: eight voices, a patch, and a buffer to fill.
    ///
    /// Nothing in here knows what is driving it. The editor drives it from a piano keyboard
    /// and a MIDI port; a plugin would drive it from a host's note events; the tests drive
    /// it from a loop and render to a file. That is deliberate - it is the one decision
    /// that keeps a plugin possible later without any of this being rewritten.
    ///
    /// THREADS
    ///
    /// Render runs on the audio thread and must never allocate, never lock and never block.
    /// Notes arrive from somewhere else, so they go into a small lock-free ring and are
    /// picked up at the top of the next Render. A lock here would be a dropout.
    /// </summary>
    public sealed class Engine
    {
        /// <summary>What the machine has. Voices past this steal the oldest.</summary>
        public const int Polyphony = 8;

        readonly Voice[] _voices = new Voice[Polyphony];
        readonly List<KeygroupPatch> _matched = new List<KeygroupPatch>(8);

        // The key ranges of whatever _matched is holding, so the crossfade can be worked out
        // without allocating on the audio thread. Sized to the polyphony because that is the
        // most voices one note can ever start.
        readonly int[] _matchedLow = new int[Polyphony];
        readonly int[] _matchedHigh = new int[Polyphony];

        // the event ring: written by whoever is playing, read by the audio thread
        struct Event { public byte Kind, A, B; }
        const int RingSize = 256;
        readonly Event[] _ring = new Event[RingSize];
        int _write, _read;

        const byte EvNoteOn = 1, EvNoteOff = 2, EvWheel = 3, EvAllOff = 4, EvRepatch = 5,
                   EvBend = 6;

        Patch _patch;
        long _sequence;

        /*
         * What the last note-on actually started.
         *
         * Not for the engine's benefit - for the caller's, so a report of "every key
         * plays the same sample" can be answered with what the voices really picked up
         * rather than with what the caller meant to ask for. References only: naming
         * them would mean building a string on the audio thread, which is the one thing
         * the render path is not allowed to do.
         */
        readonly Sound[] _started = new Sound[Polyphony];
        int _startedCount, _startedNote = -1;
        double _sharedPhase, _sharedStep;
        int _wheel;
        int _bend14 = 8192;          // the pitch wheel, at rest in the middle

        public double SampleRate { get; private set; }

        /// <summary>The master trim, as a plain gain. Eight voices at once can clip.</summary>
        public float Gain = 0.7f;

        public Engine(double sampleRate)
        {
            SampleRate = sampleRate <= 0 ? 48000 : sampleRate;
            for (int i = 0; i < _voices.Length; i++) _voices[i] = new Voice();
        }

        /// <summary>
        /// What to play. Safe to call while running: the audio thread only ever reads it,
        /// and a reference assignment is atomic.
        /// </summary>
        public void SetPatch(Patch patch)
        {
            _patch = patch;
            _sharedStep = 0;
            if (patch != null)
            {
                // The programme's own LFO runs at the rate its keygroups ask for. They
                // almost always agree; where they do not, the first one wins, since one
                // shared oscillator cannot be at two rates at once.
                for (int i = 0; i < patch.Keygroups.Count; i++)
                {
                    KeygroupPatch k = patch.Keygroups[i];
                    if (k.LfoDesync) continue;
                    _sharedStep = 2.0 * Math.PI *
                        (Cal.LfoRateHzAtZero + k.LfoRate * Cal.LfoRateHzPerUnit) / SampleRate;
                    break;
                }
            }

            // Notes already sounding take up the new settings where they stand, rather
            // than waiting to be struck again. Posted rather than done here: the voices
            // belong to the audio thread. See Repatch.
            Post(EvRepatch, 0, 0);
        }

        /// <summary>The note the last note-on was for, or -1.</summary>
        public int LastNote { get { return _startedNote; } }

        /// <summary>How many voices that note-on started.</summary>
        public int LastStartedCount { get { return _startedCount; } }

        /// <summary>The sound one of them picked up.</summary>
        public Sound LastStarted(int i)
        {
            return i >= 0 && i < _startedCount ? _started[i] : null;
        }

        public int ActiveVoices
        {
            get
            {
                int n = 0;
                for (int i = 0; i < _voices.Length; i++) if (_voices[i].Active) n++;
                return n;
            }
        }

        // ------------------------------------------------------------------ playing

        public void NoteOn(int note, int velocity) { Post(EvNoteOn, note, velocity); }
        public void NoteOff(int note) { Post(EvNoteOff, note, 0); }
        public void Modwheel(int value) { Post(EvWheel, value, 0); }

        /// <summary>
        /// The pitch wheel, 0..16383 with 8192 at rest. Split across the event's two byte
        /// fields, because fourteen bits do not fit in one.
        /// </summary>
        public void PitchBend(int value)
        {
            int v = value < 0 ? 0 : (value > 16383 ? 16383 : value);
            Post(EvBend, (v >> 7) & 0x7F, v & 0x7F);
        }

        /// <summary>
        /// How far the wheel bends, in semitones. The machine's MIDI page offers 1 to 12.
        ///
        /// Not read off a disk: it belongs to the machine rather than to a programme, and the
        /// OVERALL SETTINGS file that would hold it is written only when somebody saves it
        /// deliberately. Two is the usual default everywhere.
        /// </summary>
        public double BendRange = 2.0;
        public void AllNotesOff() { Post(EvAllOff, 0, 0); }

        void Post(byte kind, int a, int b)
        {
            int w = _write;
            int next = (w + 1) % RingSize;
            if (next == _read) return;            // full: drop it rather than block

            _ring[w].Kind = kind;
            _ring[w].A = (byte)(a < 0 ? 0 : (a > 255 ? 255 : a));
            _ring[w].B = (byte)(b < 0 ? 0 : (b > 255 ? 255 : b));
            System.Threading.Thread.MemoryBarrier();
            _write = next;
        }

        void DrainEvents()
        {
            while (_read != _write)
            {
                Event e = _ring[_read];
                _read = (_read + 1) % RingSize;

                switch (e.Kind)
                {
                    case EvNoteOn: StartNote(e.A, e.B); break;
                    case EvNoteOff: StopNote(e.A); break;
                    case EvWheel: _wheel = e.A; break;
                    case EvBend: _bend14 = (e.A << 7) | e.B; break;
                    case EvRepatch: Repatch(); break;
                    case EvAllOff:
                        for (int i = 0; i < _voices.Length; i++) _voices[i].Release();
                        break;
                }
            }
        }

        /// <summary>
        /// Hand every sounding voice its keygroup out of the patch now loaded.
        ///
        /// Runs here, on the audio thread, because the voices belong to it. SetPatch is
        /// called from whichever thread noticed the edit, and reaching into the voices
        /// from there would be a race with the block being rendered - so it posts,
        /// like a note does, and this picks it up in order with everything else.
        ///
        /// Matched by keygroup index rather than by key range: a voice belongs to the
        /// keygroup that started it, and an edit may have moved the ranges under it.
        /// </summary>
        void Repatch()
        {
            Patch p = _patch;
            if (p == null) return;

            for (int i = 0; i < _voices.Length; i++)
            {
                Voice v = _voices[i];
                if (!v.Active) continue;

                int want = v.KeygroupIndex;
                if (want < 0) continue;

                for (int k = 0; k < p.Keygroups.Count; k++)
                {
                    KeygroupPatch kg = p.Keygroups[k];
                    if (kg.KeygroupIndex != want) continue;
                    if (v.Velocity < kg.VelocityFrom || v.Velocity > kg.VelocityTo) continue;

                    v.Adopt(kg);
                    break;
                }
            }
        }

        void StartNote(int note, int velocity)
        {
            Patch p = _patch;
            _startedCount = 0;
            _startedNote = note;
            if (p == null) return;

            p.Matching(note, velocity, _matched);

            // The positional crossfade needs every keygroup answering this note at once,
            // so the ranges are gathered before any voice starts. With the flag off, or
            // with one keygroup answering, CrossfadeGain returns 1 and nothing changes.
            int matched = Math.Min(_matched.Count, _matchedLow.Length);
            for (int m = 0; m < matched; m++)
            {
                _matchedLow[m] = _matched[m].LowKey;
                _matchedHigh[m] = _matched[m].HighKey;
            }

            for (int m = 0; m < _matched.Count; m++)
            {
                KeygroupPatch kg = _matched[m];
                if (_startedCount < _started.Length) _started[_startedCount++] = kg.Sound;

                double fade = p.PositionalCrossfade && m < matched
                    ? Cal.CrossfadeGain(note, _matchedLow, _matchedHigh, matched, m)
                    : 1.0;

                // What the wheel adds, in cents. Byte 22 scales it, proportionally - the
                // machine gave 0.509 of full at byte 22 = 50, where proportional is 0.505.
                double wheelCents = Cal.LfoWheelCentsAtFull *
                                    (kg.LfoModwheelDepth / 99.0) * (_wheel / 127.0);

                if (kg.LfoDepth * Cal.LfoDepthCentsPerUnit + wheelCents < 0.5) wheelCents = 0;

                Voice v = Take();
                v.Start(kg, note, velocity, SampleRate, wheelCents, _sequence++, fade);
            }
        }

        /// <summary>
        /// Let go of every voice holding this note.
        ///
        /// Every one, not the first: pressing the same key twice before releasing it is
        /// ordinary, and leaving the older voice held would strand it.
        /// </summary>
        void StopNote(int note)
        {
            for (int i = 0; i < _voices.Length; i++)
                if (_voices[i].Active && _voices[i].Note == note && _voices[i].Held)
                    _voices[i].Release();
        }

        /// <summary>
        /// A voice to start a note on.
        ///
        /// Idle first, obviously. Then the oldest one that has already been let go and is
        /// only fading - taking that costs a tail nobody is listening to. Only if every
        /// voice is still held does it take one of those, and then the oldest.
        ///
        /// Taking the oldest regardless, which is what this did, would cut off a key the
        /// player is holding while a released note was left ringing beside it - which
        /// reads exactly like "the release did not happen".
        /// </summary>
        Voice Take()
        {
            for (int i = 0; i < _voices.Length; i++)
                if (!_voices[i].Active) return _voices[i];

            int pick = -1;
            for (int i = 0; i < _voices.Length; i++)
            {
                if (_voices[i].Held) continue;                  // still down: leave it
                if (pick < 0 || _voices[i].StartedAt < _voices[pick].StartedAt) pick = i;
            }

            if (pick < 0)
            {
                pick = 0;
                for (int i = 1; i < _voices.Length; i++)
                    if (_voices[i].StartedAt < _voices[pick].StartedAt) pick = i;
            }

            _voices[pick].Kill();
            return _voices[pick];
        }

        // ------------------------------------------------------------------- render

        /// <summary>
        /// Fill <paramref name="count"/> mono samples. Allocates nothing.
        /// </summary>
        public void Render(float[] buffer, int offset, int count)
        {
            DrainEvents();

            Array.Clear(buffer, offset, count);

            // The wheel reaches notes already sounding - that is what separates it from
            // velocity, which is settled when the key goes down.
            double bendMono = Cal.BendRatio(_bend14, BendRange);

            for (int i = 0; i < _voices.Length; i++)
                if (_voices[i].Active)
                {
                    _voices[i].SetBend(bendMono);
                    _voices[i].Render(buffer, offset, count, _sharedPhase);
                }

            _sharedPhase += _sharedStep * count;
            if (_sharedPhase > 2.0 * Math.PI) _sharedPhase %= 2.0 * Math.PI;

            float g = Gain;
            for (int i = 0; i < count; i++)
            {
                float v = buffer[offset + i] * g;
                buffer[offset + i] = v > 1f ? 1f : (v < -1f ? -1f : v);
            }
        }

        /// <summary>
        /// Fill <paramref name="count"/> stereo samples, honouring each keygroup's output port.
        ///
        /// The machine's LEFT and RIGHT sockets are two mono outputs rather than a pan pot, so
        /// a keygroup sent to one is absent from the other - 38 keygroups across four library
        /// programmes, including TUBULAR 2 whose bells read L L L L R R R R. Everything else
        /// lands on both sides at full level, which is what the mono path has always done.
        ///
        /// Allocates nothing. The mono overload above is untouched and still bit-identical.
        /// </summary>
        public void Render(float[] left, float[] right, int offset, int count)
        {
            DrainEvents();

            Array.Clear(left, offset, count);
            Array.Clear(right, offset, count);

            double bendStereo = Cal.BendRatio(_bend14, BendRange);

            for (int i = 0; i < _voices.Length; i++)
                if (_voices[i].Active)
                {
                    _voices[i].SetBend(bendStereo);
                    _voices[i].Render(left, right, offset, count, _sharedPhase);
                }

            _sharedPhase += _sharedStep * count;
            if (_sharedPhase > 2.0 * Math.PI) _sharedPhase %= 2.0 * Math.PI;

            float g = Gain;
            for (int i = 0; i < count; i++)
            {
                float l = left[offset + i] * g, r = right[offset + i] * g;
                left[offset + i]  = l > 1f ? 1f : (l < -1f ? -1f : l);
                right[offset + i] = r > 1f ? 1f : (r < -1f ? -1f : r);
            }
        }
    }
}
